NConfiguration/Serialization/SimpleTypes/Parsing/TimeSpanParser.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace NConfiguration.Serialization.SimpleTypes.Parsing
+{
+    public class TimeSpanParser
+    {
+        enum GroupName
+        {
+            Minus,
+            Day,
+            Hour,
+            Minute,
+            Second
+        }
+
+        private const RegexOptions RegularExpressionOptions =
+            RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Compiled;
+        private const string MinusPattern = "(^-)";
+        private const string PatternBase = @"(\d+(?:\.\d+)?)";
+
+        private static readonly KeyValuePair<GroupName, Func<double, TimeSpan>>[] _groupNumberToSpans =
+        {
+            new KeyValuePair<GroupName, Func<double, TimeSpan>>(GroupName.Day, TimeSpan.FromDays),
+            new KeyValuePair<GroupName, Func<double, TimeSpan>>(GroupName.Hour, TimeSpan.FromHours),
+            new KeyValuePair<GroupName, Func<double, TimeSpan>>(GroupName.Minute, TimeSpan.FromMinutes),
+            new KeyValuePair<GroupName, Func<double, TimeSpan>>(GroupName.Second, TimeSpan.FromSeconds)
+        };
+
+        private static string _buildGroupRegex(GroupName groupName, string postfix, string pattern = PatternBase)
+        {
+            return $"((?<{groupName}>{pattern}){postfix}){{0,1}}";
+        }
+
+        private static string _buildRegexp()
+        {
+            return _buildGroupRegex(GroupName.Minus, "", MinusPattern)
+                   + _buildGroupRegex(GroupName.Day, "d")
+                   + _buildGroupRegex(GroupName.Hour, "h")
+                   + _buildGroupRegex(GroupName.Minute, "m")
+                   + _buildGroupRegex(GroupName.Second, "s$");
```

Comment:
the line ending `$` does not belong to the second group. What you want to check, is that there is no leading or trailing characters which do not belong to your `Timespan` pattern.

It can be achieved by
```csharp
 return "^" +
    buildGroupRegex(GroupName.Minus, "", MinusPattern) +
    buildGroupRegex(GroupName.Day, "d") +
    // ...
    buildGroupRegex(GroupName.Second, "s") +
    "$";

```

Reply:
Thank you for comment. We don't have any group or symdols after s' group. S' group should be place in the end of line.

Reply:
notice the difference between
```regexp
1.  ((?<minute>\d+)m)?((?<second>\d+)s$)?
2. ^((?<minute>\d+)m)?((?<second>\d+)s)?$
```

(1.) will match **3m**sAA and also kj**3m**sAA
(2.) will match none of the 2 inputs above

you can test it at [regex101.com](https://regex101.com/)